File: Hard/Median Of Two Sorted Arrays/findMedianSortedArrays.py
class Solution(object):
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        len1, len2 = len(nums1), len(nums2)
        
        if len1 > len2:
            nums1, nums2, len1, len2 = nums2, nums1, len2, len1

        left, right = 0, len1
        half_len = (len1 + len2 + 1) //2
        
        while left <= right:
            divide1 = (left + right) //2
            divide2 = half_len - divide1
            max_left1 = float('-inf') if divide1 == 0 else nums1[divide1 -1]
            min_right1 = float('inf') if divide1 == len1 else nums1[divide1]
            max_left2 = float('-inf') if divide2 == 0 else nums2[divide2 -1]
            min_right2 = float('inf') if divide2 == len2 else nums2[divide2]
            if max_left1 <= min_right2 and max_left2 <= min_right1:
                if (len1 + len2) % 2==1:
                    return max(max_left1, max_left2)
                return (max(max_left1, max_left2) + min(min_right1, min_right2)) /2.0
            elif max_left1 > min_right2:
                right = divide1 -1
            else:
                left = divide1 +1

#QED
#Problem 4 (Median Of Two Sorted Arrays) - Jason Balayev (python)   

File: Hard/Median Of Two Sorted Arrays/findMedianSortedArrays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int len1 = nums1.length;
        int len2 = nums2.length;
        
        if (len1 > len2) {
            return findMedianSortedArrays(nums2,nums1);
        }
        
        int left = 0, right = len1, halfLength = (len1 + len2 +1) /2;
        while (left <= right) {
            int divide1 = (left + right) / 2;
            int divide2 = halfLength - divide1;
            int maxLeft1 = (divide1 == 0) ? Integer.MIN_VALUE : nums1[divide1 -1];
            int minRight1 = (divide1 == len1) ? Integer.MAX_VALUE : nums1[divide1];
            int maxLeft2 = (divide2 == 0) ? Integer.MIN_VALUE : nums2[divide2 -1];
            int minRight2 = (divide2 == len2) ? Integer.MAX_VALUE : nums2[divide2];
            
            if (maxLeft1 <= minRight2 && maxLeft2 <= minRight1) {
                if ((len1 + len2) % 2==1) {
                    return Math.max(maxLeft1, maxLeft2);
                }
                return (Math.max(maxLeft1, maxLeft2) + Math.min(minRight1, minRight2)) /2.0;
            } else if (maxLeft1 > minRight2) {
                right = divide1 -1; 
            } else {
                left = divide1 +1; 
            }
        }
        return 0.0;
    }
}

//QED
//Problem 4 (Median Of Two Sorted Arrays) - Jason Balayev (java)   